From 09492f2f69b2504c6e878996c7c8c08fee038b2e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?K=C3=A9vin=20L=27h=C3=B4pital?=
 <kevin.lhopital@savoirfairelinux.com>
Date: Thu, 31 Mar 2022 13:18:01 +0200
Subject: [PATCH 4/4] media/media_encoder.cpp: disable wrong pixels format
 conversion

Check if the camera output pixel format is supported by the IMX6
encoder.
If not, a conversion is done.
This can be the case with some usb cameras, which provide for low
resolutions raw data in an incompatible pixel format with the imx
encoder.
---
 src/media/media_encoder.cpp | 19 +++++--------------
 1 file changed, 5 insertions(+), 14 deletions(-)

diff --git a/src/media/media_encoder.cpp b/src/media/media_encoder.cpp
index 94ade84..6e00914 100644
--- a/src/media/media_encoder.cpp
+++ b/src/media/media_encoder.cpp
@@ -1281,22 +1281,13 @@ MediaEncoder::getHWFrame(const std::shared_ptr<VideoFrame>& input,
         }
 #elif !defined(__APPLE__)
         // Other Platforms
+        AVCodecContext* encoderCtx = encoders_[currentStreamIdx_];
         auto desc = av_pix_fmt_desc_get(static_cast<AVPixelFormat>(input->format()));
         bool isHardware = desc && (desc->flags & AV_PIX_FMT_FLAG_HWACCEL);
-        if (accel_ && accel_->isLinked() && isHardware) {
-            // Fully accelerated pipeline, skip main memory
-            output = input;
-        } else if (isHardware) {
-            // Hardware decoded frame, transfer back to main memory
-            // Transfer to GPU if we have a hardware encoder
-            // Hardware decoders decode to NV12, but Jami's supported software encoders want YUV420P
-            output = getUnlinkedHWFrame(*input.get());
-        } else if (accel_) {
-            // Software decoded frame with a hardware encoder, convert to accepted format first
-            output = getHWFrameFromSWFrame(*input.get());
-        } else {
-            output = getScaledSWFrame(*input.get());
-        }
+        if (input->format() != encoderCtx->pix_fmt)
+                output = scaler_.convertFormat(*input.get(), encoderCtx->pix_fmt);
+        else
+                output = input;
 #else
         // macOS
         output = getScaledSWFrame(*input.get());
-- 
2.25.1

